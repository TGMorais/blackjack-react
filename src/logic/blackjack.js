const blackjack = {
  BUST_SCORE: 21,
  DECK_SIZE: 52,

  /**
   * Creates a new deck of cards (somewhat ordered)
   * @return {Array} array of card objects
   */
  create: () => {
    const suits = ["spades", "diamonds", "hearts", "clubs"];
    const values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
    const specials = { 1: "Ace", 11: "Jack", 12: "Queen", 13: "King" };

    const deck = suits
      .map(s => {
        return values.map(v => {
          return {
            suit: s,
            name: specials[v] || v,
            value: v > 10 ? 10 : v
          };
        });
      })
      .reduce((acc, v) => acc.concat(v));

    return deck;
  },

  /**
   * Takes a deck, returns a new shuffled deck
   */
  shuffle: deck => {
    if (!(deck instanceof Array)) {
      throw new Error("Deck not an array");
    }

    let _deck = [...deck];
    let i = 0,
      len = _deck.length;

    const swap = (deck, a, b) => {
      let tmp = deck[a];
      deck[a] = deck[b];
      deck[b] = tmp;
    };

    while (i < len) {
      const newi = Math.floor(Math.random() * i);
      swap(_deck, i, newi);
      i++;
    }
    return _deck;
  },

  /**
   * @returns a new shuffled deck
   */
  createShuffled: () => {
    return blackjack.shuffle(blackjack.create());
  },

  /**
   * Deals numCards cards
   * @param {Array} deck - array of cards
   * @param {Number} numCards - number of cards to deal
   * @returns {{deck, hand}}  - An object with the new deck state and the cards delt
   */
  deal: (deck, numCards) => {
    //todo: asset deck = array, numCards = number

    if (numCards > deck.length) {
      throw new Error(
        `It's impossible to deal ${numCards} cards. Deck only has ${deck.length} left!`
      );
    }

    const newdeck = [...deck];
    const cards = [];
    //

    while (numCards > 0) {
      cards.push(newdeck.shift());
      numCards--;
    }

    return {
      deck: newdeck,
      cards
    };
  },

  flipCard: card => {
    return { ...card, flipped: !(card.flipped) };
  },

  //Calculates valud of hand (not counting flipped cards)
  calculateHand: hand => {
    const cardValues = hand.filter(card => !(card.flipped)).map(card => card.value)
    return blackjack.calculateHandValue(cardValues);
  },

  /**
   * @param {array} - array of card values
   * @returns {number} - value of hand 
   */
  calculateHandValue: cardValues => {
    let total = cardValues.reduce((accumulator, curr, prev) => {
      return accumulator + curr;
    }, 0);

    if (total > blackjack.BUST_SCORE) {
      let numAces = cardValues.filter(x => x === 11).length;
      while (numAces > 0 && total > blackjack.BUST_SCORE) {
        total -= 10;
        numAces--;
      }
    }

    return total;
  }
};

export default blackjack;
