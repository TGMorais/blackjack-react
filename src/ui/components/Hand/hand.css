.hand {
    border: 1px solid rgba(77, 77, 77, 0.48);
    border-radius: 10px;
    overflow: hidden;
    flex-wrap: nowrap;
    margin: 5px 0;
}
.hand .card { 
    display: inline-block;
    width: 25%;
    min-width: 50px;
    max-width: 115px;
}