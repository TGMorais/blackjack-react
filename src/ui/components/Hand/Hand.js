import React from "react";
import Card from '../Card/Card';
import './hand.css';

const Hand = ({ cards = [] }) => {
  return (
      <div className="hand flex">
          {
              /*  use index as key for now */
              cards.map((card,i) => <Card key={i} {...card} />)
          }
      </div>
  )
};

export default Hand;
