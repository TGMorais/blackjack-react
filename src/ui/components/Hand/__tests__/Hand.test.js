import React from "react";
import ReactDOM from "react-dom";
import Hand from "../Hand";
import renderer from "react-test-renderer";
//TODO: use webpack alias!!
import blackjack from "../../../../logic/blackjack";

let deck = []
beforeEach(() => {
  deck = blackjack.create();
});

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Hand />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer.create(<Hand>Button</Hand>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders with cards correctly", () => {
  const cards = [...deck.slice(0, 3)];

  const tree = renderer.create(<Hand cards={cards}>Button</Hand>).toJSON();
  expect(tree).toMatchSnapshot();
});
