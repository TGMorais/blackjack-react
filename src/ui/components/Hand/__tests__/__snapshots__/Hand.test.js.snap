// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="hand flex"
/>
`;

exports[`renders with cards correctly 1`] = `
<div
  className="hand flex"
>
  <div
    className="card "
  >
    <div
      className="suit"
    >
      ♠️
    </div>
    <div
      className="name"
    >
      A
    </div>
  </div>
  <div
    className="card "
  >
    <div
      className="suit"
    >
      ♠️
    </div>
    <div
      className="name"
    >
      2
    </div>
  </div>
  <div
    className="card "
  >
    <div
      className="suit"
    >
      ♠️
    </div>
    <div
      className="name"
    >
      3
    </div>
  </div>
</div>
`;
